opensearch-operator/pkg/reconcilers/tls.go

```diff
@@ -102,50 +105,147 @@ func (r *TLSReconciler) handleTransport() error {
+	newVersionConstraint, err := semver.NewConstraint(">=2.0.0")
+	if err != nil {
+		panic(err)
+	}
+
+	version, err := semver.NewVersion(r.instance.Spec.General.Version)
+	if err != nil {
+		r.logger.Error(err, "unable to parse version, assuming >= 2.0.0")
+		return true
+	}
+	return newVersionConstraint.Check(version)
+}
+
+func (r *TLSReconciler) adminCAName() string {
+	if r.securityChangeVersion() {
+		return r.instance.Spec.Security.Tls.Http.TlsCertificateConfig.CaSecret.Name
+	}
+	return r.instance.Spec.Security.Tls.Transport.TlsCertificateConfig.CaSecret.Name
+}
+
+func (r *TLSReconciler) adminCAProvided() bool {
+	return r.adminCAName() != ""
+}
+
+func (r *TLSReconciler) providedCAForAdminCert() (tls.Cert, error) {
+	return r.providedCaCert(
+		r.adminCAName(),
+		r.instance.Namespace,
+	)
+}
+
+func (r *TLSReconciler) getCACert() (tls.Cert, error) {
+	if r.adminCAProvided() {
+		return r.providedCAForAdminCert()
+	}
+	return util.ReadOrGenerateCaCert(r.pki, r.Client, r.ctx, r.instance)
+}
+
+func (r *TLSReconciler) shouldCreateAdminCert(ca tls.Cert) (bool, error) {
+	secret := &corev1.Secret{}
+	err := r.Get(r.ctx, types.NamespacedName{
+		Name:      r.adminSecretName(),
+		Namespace: r.instance.Namespace,
+	}, secret)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.logger.Info("admin cert does not exist, creating")
+			return true, nil
+		}
+		return false, err
+	}
+
+	data, ok := secret.Data[corev1.TLSCertKey]
+	if !ok {
+		return true, nil
+	}
+
+	validator, err := tls.NewCertValidater(data, tls.WithExpiryThreshold(5*24*time.Hour))
+	if err != nil {
+		return false, err
+	}
+
+	if validator.IsExpiringSoon() {
+		r.logger.Info("admin cert is expiring soon, recreating")
+		return true, nil
+	}
+
+	verified, err := validator.IsSignedByCA(ca)
+	if err != nil {
+		return false, err
+	}
+
+	if !verified {
+		r.logger.Info("admin cert is not signed by CA, recreating")
+	}
+
+	return !verified, nil
+}
+
+func (r *TLSReconciler) createAdminSecret(ca tls.Cert) error {
+	createCert, err := r.shouldCreateAdminCert(ca)
+	if err != nil {
+		return fmt.Errorf("failed to determine if admin cert should be created: %w", err)
+	}
+	if !createCert {
+		return nil
+	}
+
+	adminCert, err := ca.CreateAndSignCertificate("admin", r.instance.Name, nil)
+	if err != nil {
+		r.logger.Error(err, "Failed to create admin certificate", "interface", "transport")
+		r.recorder.AnnotatedEventf(
+			r.instance,
+			map[string]string{"cluster-name": r.instance.GetName()},
+			"Warning",
+			"Security",
+			"Failed to create admin certificate",
+		)
+		return err
+	}
+	adminSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.adminSecretName(),
+			Namespace: r.instance.Namespace,
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: adminCert.SecretData(ca),
+	}
+	if err := ctrl.SetControllerReference(r.instance, adminSecret, r.Client.Scheme()); err != nil {
+		return err
+	}
+	return client.IgnoreAlreadyExists(r.Create(r.ctx, adminSecret))
```

Comment:
Shouldn't this be create or update (e.g. in cases where the cert exists but is expired)? Maybe best to use `r.ReconileResource`?

Reply:
That's a good point, I will update it.